mico-core/src/main/java/io/github/ust/mico/core/web/ApplicationController.java

```diff
@@ -184,46 +208,62 @@
     public ResponseEntity<Resources<Resource<MicoService>>> getServicesFromApplication(@PathVariable(PATH_VARIABLE_SHORT_NAME) String shortName,
                                                                                        @PathVariable(PATH_VARIABLE_VERSION) String version) {
         MicoApplication micoApplication = getApplicationFromDatabase(shortName, version);
-        List<MicoService> micoServices = micoApplication.getServices();
+        List<MicoService> micoServices = serviceRepository.findAllByApplication(micoApplication.getShortName(), micoApplication.getVersion());
         List<Resource<MicoService>> micoServicesWithLinks = ServiceController.getServiceResourcesList(micoServices);
         return ResponseEntity.ok(
             new Resources<>(micoServicesWithLinks,
                 linkTo(methodOn(ApplicationController.class).getServicesFromApplication(shortName, version)).withSelfRel()));
     }
 
+    // TODO: Deployment info necessary to add new service to application!
     @PostMapping("/{" + PATH_VARIABLE_SHORT_NAME + "}/{" + PATH_VARIABLE_VERSION + "}/" + PATH_SERVICES)
     public ResponseEntity<Void> addServiceToApplication(@PathVariable(PATH_VARIABLE_SHORT_NAME) String applicationShortName,
                                                         @PathVariable(PATH_VARIABLE_VERSION) String applicationVersion,
                                                         @RequestBody MicoService providedService) {
         MicoApplication application = getApplicationFromDatabase(applicationShortName, applicationVersion);
         MicoService existingService = validateProvidedService(providedService);
-
-        if (!application.getServices().contains(existingService)) {
-            log.info("Add service '" + existingService.getShortName() + "' '" + existingService.getVersion() +
-                "' to application '" + applicationShortName + "' '" + applicationVersion + "'.");
-            application.getServices().add(existingService);
+        
+        // Each service can be added to one application only once
+        if (!application.getServiceDeploymentInfos().stream().anyMatch(sdi -> sdi.getService().equals(existingService))) {
```

Comment:
Condition can be simplified: `application.getServiceDeploymentInfos().stream().noneMatch`

Reply:
Done.